SpiceSharp/Sparse/SparseUtilities.cs

```diff
@@ -0,0 +1,323 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace SpiceSharp.Sparse
+{
+    /// <summary>
+    /// Utility functions for sparse matrices
+    /// </summary>
+    public static class SparseUtilities
+    {
+        /// <summary>
+        /// Preordering for Modified Nodal Analysis
+        /// </summary>
+        /// <param name="matrix">Matrix</param>
+        public static void PreorderMNA(this Matrix matrix)
+        {
+            int J, Size;
+            MatrixElement pTwin1 = null, pTwin2 = null;
+            int Twins, StartAt = 1;
+            bool Swapped, AnotherPassNeeded;
+
+            // Begin `spMNA_Preorder'. 
+            if (matrix.Factored)
+                throw new SparseException("Matrix is factored");
+
+            if (matrix.RowsLinked)
+                return;
+            Size = matrix.IntSize;
+            matrix.Reordered = true;
+
+            do
+            {
+                AnotherPassNeeded = Swapped = false;
+
+                // Search for zero diagonals with lone twins. 
+                for (J = StartAt; J <= Size; J++)
+                {
+                    if (matrix.Diag[J] == null)
+                    {
+                        Twins = CountTwins(matrix, J, ref pTwin1, ref pTwin2);
+                        if (Twins == 1)
+                        {
+                            // Lone twins found, swap rows. 
+                            SwapCols(matrix, pTwin1, pTwin2);
+                            Swapped = true;
+                        }
+                        else if ((Twins > 1) && !AnotherPassNeeded)
+                        {
+                            AnotherPassNeeded = true;
+                            StartAt = J;
+                        }
+                    }
+                }
+
+                // All lone twins are gone, look for zero diagonals with multiple twins. 
+                if (AnotherPassNeeded)
+                {
+                    for (J = StartAt; !Swapped && (J <= Size); J++)
+                    {
+                        if (matrix.Diag[J] == null)
+                        {
+                            Twins = CountTwins(matrix, J, ref pTwin1, ref pTwin2);
```

Comment:
@svenboulanger We don't use Twins in later code. Why do we need to compute it

Reply:
Remnant from Spice 3f5. Removed.